lighthouse-plugin-publisher-ads/audits/viewport-ad-density.js

```diff
@@ -64,22 +110,23 @@ class ViewportAdDensity extends Audit {
       return auditNotApplicable.NoVisibleSlots;
     }
 
-    const adArea = slots.reduce((sum, slot) =>
-      sum + boxViewableArea(slot.clientRect, viewport), 0);
-
-    const viewArea = viewport.innerWidth * viewport.innerHeight;
-
-    if (viewArea <= 0) {
+    if (viewport.innerHeight <= 0) {
       throw new Error(auditError.ViewportAreaZero);
     }
-    if (adArea > viewArea) {
-      throw new Error(auditError.AreaLargerThanViewport);
-    }
-    const adDensity = adArea / viewArea;
+
+    const adsLength = computeAdLength(slots, viewport);
+
+    // We measure document length based on the bottom ad so that it isn't skewed
+    // by lazy loading.
+    const adsBottom =
+      Math.max(...slots.map((s) => s.clientRect.top + s.clientRect.height / 2));
```

Comment:
Why divide height by 2 here? Wouldn't be want to be looking at the bottom-most pixel, the the middle one? In which case you could just use `clientRect.bottom`.

Reply:
That's a bit tricky. I was initially thinking about doing the bottom ad top, but that makes the ad size a larger factor whether this audit fails which isn't good for determinism. I thought about bottom ad bottom, but then we might fail to capture lazy loads where the ads are ~1 viewport apart since the lower ad wouldn't be in the viewport. So I split the difference and chose the middle. Plus 50% in the viewport is the viewability standard for being on screen.